### ISSUE TITLE
RBAC on GKE - extra step needed

### ISSUE THREAD

**Issue by gytisgreitai** [coreos/prometheus-operator#357]
**What did you do?**

`hack/cluster-monitoring/deploy`

**What did you expect to see?**

No errors

**What did you see instead? Under which circumstances?**

```
namespace "monitoring" created
clusterrolebinding "prometheus-operator" configured
serviceaccount "prometheus-operator" created
deployment "prometheus-operator" created
Error from server (Forbidden): error when creating "manifests/prometheus-operator/prometheus-operator-cluster-role.yaml": clusterroles.rbac.authorization.k8s.io "prometheus-operator" is forbidden: attempt to grant extra privileges: [{[create] [extensions] [thirdpartyresources] [] []} {[*] [monitoring.coreos.com] [alertmanagers] [] []} {[*] [monitoring.coreos.com] [prometheuses] [] []} {[*] [monitoring.coreos.com] [servicemonitors] [] []} {[*] [apps] [statefulsets] [] []} {[*] [] [configmaps] [] []} {[*] [] [secrets] [] []} {[list] [] [pods] [] []} {[delete] [] [pods] [] []} {[get] [] [services] [] []} {[create] [] [services] [] []} {[update] [] [services] [] []} {[get] [] [endpoints] [] []} {[create] [] [endpoints] [] []} {[update] [] [endpoints] [] []} {[list] [] [nodes] [] []} {[watch] [] [nodes] [] []}] user=&{gytis@tripcreator.com  [system:authenticated] map[]} ownerrules=[{[create] [authorization.k8s.io] [selfsubjectaccessreviews] [] []} {[get] [] [] [] [/api /api/* /apis /apis/* /healthz /swaggerapi /swaggerapi/* /version]}] ruleResolutionErrors=[]
Waiting for Operator to register third party objects...done!
daemonset "node-exporter" created
service "node-exporter" created
clusterrolebinding "kube-state-metrics" configured
deployment "kube-state-metrics" created
serviceaccount "kube-state-metrics" created
service "kube-state-metrics" created
Error from server (Forbidden): error when creating "manifests/kube-state-metrics/kube-state-metrics-cluster-role.yaml": clusterroles.rbac.authorization.k8s.io "kube-state-metrics" is forbidden: attempt to grant extra privileges: [{[list] [] [nodes] [] []} {[watch] [] [nodes] [] []} {[list] [] [pods] [] []} {[watch] [] [pods] [] []} {[list] [] [resourcequotas] [] []} {[watch] [] [resourcequotas] [] []} {[list] [extensions] [daemonsets] [] []} {[watch] [extensions] [daemonsets] [] []} {[list] [extensions] [deployments] [] []} {[watch] [extensions] [deployments] [] []} {[list] [extensions] [replicasets] [] []} {[watch] [extensions] [replicasets] [] []}] user=&{gytis@tripcreator.com  [system:authenticated] map[]} ownerrules=[{[create] [authorization.k8s.io] [selfsubjectaccessreviews] [] []} {[get] [] [] [] [/api /api/* /apis /apis/* /healthz /swaggerapi /swaggerapi/* /version]}] ruleResolutionErrors=[]
secret "grafana-credentials" created
secret "grafana-credentials" configured
configmap "grafana-dashboards" created
deployment "grafana" created
service "grafana" created
clusterrolebinding "prometheus" configured
configmap "prometheus-k8s-rules" created
serviceaccount "prometheus-k8s" created
servicemonitor "alertmanager" configured
servicemonitor "kube-apiserver" configured
servicemonitor "k8s-apps-http" configured
servicemonitor "kube-state-metrics" configured
servicemonitor "kubelet" configured
servicemonitor "node-exporter" configured
servicemonitor "prometheus" configured
service "prometheus-k8s" created
prometheus "k8s" configured
Error from server (Forbidden): error when creating "manifests/prometheus/prometheus-cluster-role.yaml": clusterroles.rbac.authorization.k8s.io "prometheus" is forbidden: attempt to grant extra privileges: [{[get] [] [nodes] [] []} {[list] [] [nodes] [] []} {[watch] [] [nodes] [] []} {[get] [] [services] [] []} {[list] [] [services] [] []} {[watch] [] [services] [] []} {[get] [] [endpoints] [] []} {[list] [] [endpoints] [] []} {[watch] [] [endpoints] [] []} {[get] [] [pods] [] []} {[list] [] [pods] [] []} {[watch] [] [pods] [] []} {[get] [] [configmaps] [] []} {[get] [] [] [] [/metrics]}] user=&{gytis@tripcreator.com  [system:authenticated] map[]} ownerrules=[{[create] [authorization.k8s.io] [selfsubjectaccessreviews] [] []} {[get] [] [] [] [/api /api/* /apis /apis/* /healthz /swaggerapi /swaggerapi/* /version]}] ruleResolutionErrors=[]
secret "alertmanager-main" created
service "alertmanager-main" created
alertmanager "main" configured
```

**Environment**

* Kubernetes version information:

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:33:11Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:22:08Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 

	GKE

According to Google Container Engine docs:

> 
> Because of the way Container Engine checks permissions when you create a Role or ClusterRole, you must first create a RoleBinding that grants you all of the permissions included in the role you want to create.
> 
> An example workaround is to create a RoleBinding that gives your Google identity a cluster-admin role before attempting to create additional Role or ClusterRole permissions.
> 
> This is a known issue in the Beta release of Role-Based Access Control in Kubernetes and Container Engine version 1.6.
> 

So in order to proceed without error, cluster-admin role should be added to current executing user, eg:

```kubectl create clusterrolebinding your-user-cluster-admin-binding --clusterrole=cluster-admin --user=your.google.cloud.email@example.org```

Could this be added as a faq/hint etc. somewhere?

Thank you!

**Comment by brancz** [comment]
Yes I think it would definitely be worth starting an FAQ/Troubleshooting document in the Documentation folder. Do you want to do a PR for this @gytisgreitai ?

**Comment by gytisgreitai** [comment]
sure :)

**Comment by gytisgreitai** [comment]
Created https://github.com/coreos/prometheus-operator/pull/360

**Comment by brancz** [comment]
Closing here as #360 is merged. Thanks a lot @gytisgreitai !

**Comment by jason-riddle** [comment]
So I followed the instructions in #360, but I'm still stuck with "forbidden" access. Is there anything else need to be done? Do the credentials need to be recreated or do any other commands need to be ran?

**Comment by gytisgreitai** [comment]
Should be fine just with that command.

- What Kubernetes version are you using?
- Do you see your clusterrolebinding when doing `kubectl get clusterrolebinding` ?
- What does `kubectl get clusterrolebinding [yourclusterindingname] -o json` output?

As a quick hack you can do `kubectl proxy` and apply those failed yamls via `create`

**Comment by jason-riddle** [comment]
- `kubectl version`

```
$ kubectl version --short
Client Version: v1.6.0
Server Version: v1.6.2
```

- I do see my clusterrolebinding.

```
$ kubectl get clusterrolebinding | grep myname-cluster-admin-binding
myname-cluster-admin-binding                   17h
```

- `kubectl get clusterrolebinding myname-cluster-admin-binding -o json`

```json
{
    "apiVersion": "rbac.authorization.k8s.io/v1beta1",
    "kind": "ClusterRoleBinding",
    "metadata": {
        "creationTimestamp": "2017-05-14T23:09:25Z",
        "name": "myname-cluster-admin-binding",
        "resourceVersion": "188788",
        "selfLink": "/apis/rbac.authorization.k8s.io/v1beta1/clusterrolebindingsmyname-cluster-admin-binding",
        "uid": "<snipped>"
    },
    "roleRef": {
        "apiGroup": "rbac.authorization.k8s.io",
        "kind": "ClusterRole",
        "name": "cluster-admin"
    },
    "subjects": [
        {
            "apiGroup": "rbac.authorization.k8s.io",
            "kind": "User",
            "name": "<my-gmail-username>@gmail.com"
        }
    ]
}
```

I literally used the name `myname-cluster-admin-binding`, but that shouldn't be an issue correct?

**Comment by gytisgreitai** [comment]
Well I've just tested this on 1.6.2 and it seems to work fine for me.
And `gcloud info | grep Account` outputs `<my-gmail-username>@gmail.com` yes?

What access rights does this email have on iAM ?

**Comment by jason-riddle** [comment]
- Yes, `gcloud info | grep Account` outputs my email.
- This user is an Owner on the project.

![screen shot 2017-05-16 at 9 59 28 am](https://cloud.githubusercontent.com/assets/2156866/26110120/54011580-3a1f-11e7-97d2-102b9c079cd6.jpg)

I'll let this issue die out since it only seems to be my problem, but I'll report back if I find a fix.

**Comment by ngtuna** [comment]
> Yes, gcloud info | grep Account outputs my email.

That's not always true. For example, the case of @yuvipanda showed that the username that GKE recognized was capitalized at some characters. So my two questions:

1) why does capitalization happen ?
2) How do we retrieve that _modified/capitalized/real_ username by code ? I am looking for support from client-go for example...

**Comment by ngtuna** [comment]
Okay I found this below command actually better:

```
$ gcloud projects get-iam-policy PROJECT_ID
```
It returns the user's email and in my case the capitalized one. So, probably it's an alternative command if `gcloud info | grep Account` doesn't work.

**Comment by peterkuiper** [comment]
Tried all the above, but still no luck:

```
kubectl version --short
Client Version: v1.8.4
Server Version: v1.8.4-gke.0
```
```
Error: release prometheus-operator failed: clusterroles.rbac.authorization.k8s.io "prometheus-operator-prometheus-operator" is forbidden: attempt to grant extra privileges
```

```
➜  prometheus-operator git:(master) git rev-parse HEAD
848497cb6dfbab44f99c8facc2da5c93e87ec6c4
```

Anything else I could try?

**Comment by brancz** [comment]
That's very curious, as this seems to continue to be an RBAC issue and (so far) has only ever been reported for GKE, I would try to contact their support.

**Comment by mattnworb** [comment]
For what it is worth, `gcloud auth list` and `gcloud config get-value account` might be an easier way to check the currently configured account name.

**Comment by brancz** [comment]
@mattnworb feel free to open a PR to improve docs where possible :slightly_smiling_face:

**Comment by jason-riddle** [comment]
Based on @ngtuna's response, here's what I did to streamline the process a bit further.

```bash
PROJECT_ID='<project_id>'

gcloud projects get-iam-policy "$PROJECT_ID" --format json \
  | jq -r '.bindings[] | select(.role == "roles/owner") | .members[]' \
  | awk -F':' '{print $2}'

# CapitalizedEmail@gmail.com
```

@mattnworb Both `gcloud auth list` and `gcloud config get-value account` still returned the incorrect un-capitalized email address for me.

**Comment by philicious** [comment]
@jason-riddle your code snippet is nice but only working for roles/owner. it can happen that the user is a `roles/container.clusterAdmin` but not a project owner. 
`gcloud auth list` and `gcloud config get-value account` also return the all lowercase user for me

**Comment by Blasterdick** [comment]
Instructions from [here](https://coreos.com/operators/prometheus/docs/latest/troubleshooting.html) seem to doesn't work, as I have `clusterrolebinding` with my G Suite account email pointed, but no use. 

```

**Comment by Blasterdick** [comment]
Sorted this out. Go to [Stackdriver Logging](https://console.cloud.google.com/logs/viewer), select appropriate kubernetes cluster and error log level. 

Apply the next advanced filter:

```
resource.type="k8s_cluster"
resource.labels.location="europe-west1-b"
resource.labels.cluster_name="your-cluster-name"
severity>=ERROR
protoPayload.resourceName="rbac.authorization.k8s.io/v1beta1/clusterroles/prometheus-operator"
```
And you'll find errors like:

```

**Comment by RobinUS2** [comment]
@Blasterdick it may be obvious but keep in mind that using an actual email address for a user account e.g. john@doe.com will also work

**Comment by Blasterdick** [comment]
@RobinUS2 for my setup with GKE v1.9.2 cluster using an actual G Suite account email doesn't work, I've tried it first.

**Comment by RobinUS2** [comment]
@Blasterdick interesting, because in my case there was no principal ID number to be found and I tried the email and it did work. At least it's worth testing both in case you are not making progress.

**Comment by fracasula** [comment]
What I needed in my case was to create a [ClusterRoleBinding](https://kubernetes.io/docs/admin/authorization/rbac/#rolebinding-and-clusterrolebinding) to give my helm tiller super-powers :muscle: 

First of all:

```bash
kubectl create serviceaccount tiller --namespace kube-system
```

Then create a `tiller-clusterrolebinding.yaml` file:

```yaml
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: tiller-clusterrolebinding
subjects:
- kind: ServiceAccount
  name: tiller
  namespace: kube-system
roleRef:
  kind: ClusterRole
  name: cluster-admin
  apiGroup: ""
```

And finally:

```bash
kubectl create -f tiller-clusterrolebinding.yaml
helm init --service-account tiller --upgrade
```

[Source](https://docs.bitnami.com/kubernetes/how-to/configure-rbac-in-your-kubernetes-cluster/).

**Comment by cakirilker** [comment]
@fracasula I followed your instructions and still getting "forbidden" output as detailed below.
```javascript
function fancyAlert(arg) {
  if(arg) {
    $.facebox({div:'#foo'})
  }
}
```

**Comment by cakirilker** [comment]
@fracasula  I've followed your instructions and still getting "forbidden" output as detailed below:

> Error: release prometheus-operator failed: clusterroles.rbac.authorization.k8s.io "prometheus-operator" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["thirdpartyresources"], APIGroups:["extensions"], Verbs:["*"]} PolicyRule{Resources:["customresourcedefinitions"], APIGroups:["apiextensions.k8s.io"], Verbs:["*"]} PolicyRule{Resources:["alertmanager"], APIGroups:["monitoring.coreos.com"], Verbs:["*"]} PolicyRule{Resources:["alertmanagers"], APIGroups:["monitoring.coreos.com"], Verbs:["*"]} PolicyRule{Resources:["prometheus"], APIGroups:["monitoring.coreos.com"], Verbs:["*"]} PolicyRule{Resources:["prometheuses"], APIGroups:["monitoring.coreos.com"], Verbs:["*"]} PolicyRule{Resources:["service-monitor"], APIGroups:["monitoring.coreos.com"], Verbs:["*"]} PolicyRule{Resources:["servicemonitors"], APIGroups:["monitoring.coreos.com"], Verbs:["*"]} PolicyRule{Resources:["statefulsets"], APIGroups:["apps"], Verbs:["*"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["*"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["*"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["delete"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["create"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["update"]} PolicyRule{Resources:["endpoints"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["endpoints"], APIGroups:[""], Verbs:["create"]} PolicyRule{Resources:["endpoints"], APIGroups:[""], Verbs:["update"]} PolicyRule{Resources:["nodes"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["nodes"], APIGroups:[""], Verbs:["watch"]} PolicyRule{Resources:["namespaces"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["namespaces"], APIGroups:[""], Verbs:["watch"]}] user=&{system:serviceaccount:kube-system:tiller 7b43a000-51fe-11e8-9350-00155dd8f11a [system:serviceaccounts system:serviceaccounts:kube-system system:authenticated] map[]} ownerrules=[] ruleResolutionErrors=[[clusterroles.rbac.authorization.k8s.io "cluster-admin" not found, clusterroles.rbac.authorization.k8s.io "cluster-admin" not found]]

I'm using minikube v0.25.0 on my local Windows machine.

````
kubectl get clusterrolebindings
tiller                      1h
tiller-clusterrolebinding   6m
````
````yaml
kubectl get clusterrolebinding -o=yaml

apiVersion: v1
items:
- apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    creationTimestamp: 2018-05-08T18:40:07Z
    name: tiller
    namespace: ""
    resourceVersion: "38831"
    selfLink: /apis/rbac.authorization.k8s.io/v1/clusterrolebindings/tiller
    uid: 3ad20aa3-52ef-11e8-9e50-00155dd8f11a
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: cluster-admin
  subjects:
  - kind: ServiceAccount
    name: tiller
    namespace: kube-system
- apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    creationTimestamp: 2018-05-08T19:52:43Z
    name: tiller-clusterrolebinding
    namespace: ""
    resourceVersion: "41447"
    selfLink: /apis/rbac.authorization.k8s.io/v1/clusterrolebindings/tiller-clusterrolebinding
    uid: 5f8b5e03-52f9-11e8-a703-00155dd8f11a
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: cluster-admin
  subjects:
  - kind: ServiceAccount
    name: tiller
    namespace: kube-system
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
````
````
kubectl version --short
Client Version: v1.9.7
Server Version: v1.9.0
````

**Comment by sipian** [comment]
To add to Blasterdick [solution](https://github.com/coreos/prometheus-operator/issues/357#issuecomment-363498455) of looking through Stackdriver logs.
We can also get the principalEmail from the `client-id` field in service-account.json file

**Comment by zhenyulin** [comment]
followed the solutions here adding both capitalised and lower-case email account to clusterrolebindings, still getting 403, has anyone else run into the same situation?

**Comment by BenoitAverty** [comment]
I also had to use the user ID in my ClusterRoleBinding instead of the email address.
see @sipian post for how to find it.

**Comment by hpurmann** [comment]
I just ran into the problem that `gcloud config get-value account` returned my email as `@googlemail.com` whereas kubernetes was expecting my account as `@gmail.com`. So assigning a cluster role binding to the `@gmail` user was required for it to work. Maybe this helps someone.

**Comment by mirthy** [comment]
Got to here because of the forbidden attempt to grant extra privileges, the easiest fix is to upgrade the cluster 1.12.x if you don't want to add a clusterrolebinding - https://cloud.google.com/kubernetes-engine/docs/how-to/role-based-access-control#iam-rolebinding-bootstrap 

Google Cloud IAM and 1.12.x does allow the creation of Roles and ClusterRoles
